File: redtra_customisation/override/purchase_invoice_point_adjustment.py
# Copyright (c) 2026, redtra_customisation contributors

import frappe
from frappe import _
from frappe.utils import cint, flt

import erpnext


def sync_point_adjustment_rows(doc):
	item_by_idx = {cint(row.idx): row for row in doc.get("items") or []}
	item_by_name = {row.name: row for row in doc.get("items") or []}

	for adj in doc.get("point_adjustments") or []:
		item = None
		if adj.purchase_invoice_item:
			item = item_by_name.get(adj.purchase_invoice_item)
		if not item and adj.item_row:
			item = item_by_idx.get(cint(adj.item_row))

		if not item:
			frappe.throw(
				_("Point Adjustment row {0}: select a valid item line.").format(adj.idx)
			)

		adj.purchase_invoice_item = item.name
		adj.item_row = item.idx
		adj.item_code = item.item_code
		adj.item_name = item.item_name
		adj.reference_amount = flt(item.net_amount)


def apply_point_adjustments(doc):
	adjustments = doc.get("point_adjustments") or []
	adjustment_by_item = {}

	for adj in adjustments:
		if not adj.purchase_invoice_item:
			continue
		adjustment_by_item.setdefault(adj.purchase_invoice_item, 0.0)
		adjustment_by_item[adj.purchase_invoice_item] += flt(adj.adjustment_amount)

	total_adjustment = 0.0
	base_total_adjustment = 0.0
	conversion_rate = flt(doc.conversion_rate) or 1.0

	for item in doc.get("items") or []:
		item_adjustment = flt(adjustment_by_item.get(item.name, 0))
		item.custom_point_adjustment_total = item_adjustment

		if item_adjustment:
			base_adjustment = flt(
				item_adjustment * conversion_rate, item.precision("base_net_amount")
			)
			item.net_amount = flt(item.net_amount) + item_adjustment
			item.base_net_amount = flt(item.base_net_amount) + base_adjustment
			total_adjustment += item_adjustment
			base_total_adjustment += base_adjustment

		item.custom_adjusted_net_amount = flt(item.net_amount)

	doc.custom_total_point_adjustment = total_adjustment

	if not total_adjustment:
		return

	doc.net_total = flt(doc.net_total) + total_adjustment
	doc.base_net_total = flt(doc.base_net_total) + base_total_adjustment
	doc.grand_total = flt(doc.grand_total) + total_adjustment
	doc.base_grand_total = flt(doc.base_grand_total) + base_total_adjustment

	if doc.meta.get_field("outstanding_amount") and doc.docstatus == 0:
		paid_amount = flt(doc.paid_amount)
		doc.outstanding_amount = flt(doc.grand_total) - paid_amount

	if doc.get("disable_rounded_total"):
		return

	if not doc.meta.get_field("rounded_total") or doc.is_rounded_total_disabled():
		return

	from frappe.utils.data import round_based_on_smallest_currency_fraction

	doc.rounded_total = round_based_on_smallest_currency_fraction(
		doc.grand_total, doc.currency, doc.precision("rounded_total")
	)
	doc.rounding_adjustment = flt(
		doc.rounded_total - doc.grand_total, doc.precision("rounding_adjustment")
	)
	if doc.meta.get_field("base_rounded_total"):
		doc.base_rounded_total = round_based_on_smallest_currency_fraction(
			doc.base_grand_total, erpnext.get_company_currency(doc.company), doc.precision("base_rounded_total")
		)
		doc.base_rounding_adjustment = flt(
			doc.base_rounded_total - doc.base_grand_total, doc.precision("base_rounding_adjustment")
		)

File: redtra_customisation/public/js/purchase_invoice_point_adjustment.js
// Copyright (c) 2026, redtra_customisation contributors

function get_pi_item_by_row(frm, item_row) {
	return (frm.doc.items || []).find(row => cint(row.idx) === cint(item_row));
}

function sync_point_adjustment_row(frm, cdt, cdn) {
	const adj = locals[cdt][cdn];
	if (!adj.item_row) {
		return;
	}

	const item = get_pi_item_by_row(frm, adj.item_row);
	if (!item) {
		frappe.msgprint(
			__("Item Row #{0} was not found. Add the item line first.", [adj.item_row])
		);
		return;
	}

	frappe.model.set_value(cdt, cdn, "purchase_invoice_item", item.name);
	frappe.model.set_value(cdt, cdn, "item_code", item.item_code);
	frappe.model.set_value(cdt, cdn, "item_name", item.item_name);
	frappe.model.set_value(cdt, cdn, "reference_amount", flt(item.net_amount));
}

function recalculate_pi_with_point_adjustments(frm) {
	if (!frm.doc.point_adjustments?.length) {
		return;
	}
	frm.script_manager.trigger("calculate_taxes_and_totals");
}

function open_point_adjustment_dialog(frm, item_row) {
	const item = get_pi_item_by_row(frm, item_row);
	if (!item) {
		frappe.msgprint(__("Select a valid item line first."));
		return;
	}

	const d = new frappe.ui.Dialog({
		title: __("Point Adjustment — Row {0}", [item.idx]),
		fields: [
			{
				fieldtype: "Data",
				fieldname: "item_label",
				label: __("Item"),
				read_only: 1,
				default: `${item.item_code || ""} — ${item.item_name || ""}`.trim(),
			},
			{
				fieldtype: "Currency",
				fieldname: "reference_amount",
				label: __("Current Line Amount"),
				read_only: 1,
				default: flt(item.net_amount),
			},
			{
				fieldtype: "Currency",
				fieldname: "adjustment_amount",
				label: __("Adjustment Amount"),
				reqd: 1,
				description: __("Use positive or negative values. Example: 0.48 on a 4.98 line."),
			},
			{
				fieldtype: "Data",
				fieldname: "remarks",
				label: __("Remarks"),
			},
		],
		primary_action_label: __("Add Adjustment"),
		primary_action(values) {
			if (!flt(values.adjustment_amount)) {
				frappe.msgprint(__("Enter a non-zero adjustment amount."));
				return;
			}

			const child = frm.add_child("point_adjustments");
			child.item_row = item.idx;
			child.purchase_invoice_item = item.name;
			child.item_code = item.item_code;
			child.item_name = item.item_name;
			child.reference_amount = flt(item.net_amount);
			child.adjustment_amount = flt(values.adjustment_amount);
			child.remarks = values.remarks;
			frm.refresh_field("point_adjustments");
			recalculate_pi_with_point_adjustments(frm);
			d.hide();
		},
	});
	d.show();
}

frappe.ui.form.on("Purchase Invoice", {
	refresh(frm) {
		if (frm.doc.docstatus !== 0) {
			return;
		}

		frm.fields_dict.items?.grid?.add_custom_button(
			__("Point Adjustment"),
			() => {
				const selected = frm.fields_dict.items.grid.get_selected();
				if (selected?.length === 1) {
					open_point_adjustment_dialog(frm, locals[selected[0].doctype][selected[0].name].idx);
					return;
				}
				frappe.msgprint(__("Select one item row, then click Point Adjustment."));
			},
			__("Adjust")
		);
	},
	items_remove(frm) {
		recalculate_pi_with_point_adjustments(frm);
	},
});

frappe.ui.form.on("Purchase Invoice Point Adjustment", {
	item_row(frm, cdt, cdn) {
		sync_point_adjustment_row(frm, cdt, cdn);
		recalculate_pi_with_point_adjustments(frm);
	},
	adjustment_amount(frm, cdt, cdn) {
		recalculate_pi_with_point_adjustments(frm);
	},
	point_adjustments_remove(frm) {
		recalculate_pi_with_point_adjustments(frm);
	},
});
